ext/winevt/winevt_subscribe.c

```diff
@@ -355,6 +357,13 @@ rb_winevt_subscribe_next(VALUE self)
     return Qfalse;
   }
 
+  dwWait = WaitForSingleObject(winevtSubscribe->signalEvent, 0);
```

Comment:
Because the document about [EvtSubscribe](https://learn.microsoft.com/en-us/windows/win32/api/winevt/nf-winevt-evtsubscribe) doesn't describe the detail of SignalEvent, I'm not sure it's correct or not.

I'm wondering the time out value `0` is correct or not.
It means it doesn't wait, return immediately and don't block: https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitforsingleobject

Reply:
Does this aim to just confirm the signal state at that time (There is no need to wait)?

Reply:
For enumerator, it is unnatural for blocking use case for WaitForSingleObject. However, this approach is not fully fulfilled for the detecting failure of subscriptions. I'm closing this for now.

Reply:
Ah, my understanding is incorrect. INFINITE waiting should be needed here. So, I rewrote this line.

Reply:
When using INFINITE timeout here, it will be waiting forever even if SIGINT is caught in the parent process of Fluentd. It's sick behavior for us.

Reply:
> This implementation is inspired from the article: https://learn.microsoft.com/ja-jp/windows/win32/wes/subscribing-to-events#pull-subscriptions EvtSubscribe can tell internal state with a registered singal event.

That's the what I wanted to know, thanks! Could you add this link as a comment or commit message?

> When using INFINITE timeout here, it will be waiting forever even if SIGINT is caught in the parent process of Fluentd. It's sick behavior for us.

I agree with you, I didn't intend to use `INFINITE` here.
But how about using small non zero value (e.g. 100) instead?

Reply:
> > When using INFINITE timeout here, it will be waiting forever even if SIGINT is caught in the parent process of Fluentd. It's sick behavior for us.
> 
> I agree with you, I didn't intend to use `INFINITE` here. But how about using small non zero value (e.g. 100) instead?

Hmm..., there is a discussion for this waiting period. I didn't want to wait for additional time here. Like Ruby's yield, it should be returned immediately to the callee. 

For implementation perspective, adding waiting period should degrade loop performance here:

Reply:
Is this signal updated in another thread?
Do we have to wait for the signal here?

Reply:
The signal is updated with EvtSubscribe's internal mechanism. Could you read [the article](https://learn.microsoft.com/ja-jp/windows/win32/wes/subscribing-to-events#pull-subscriptions) carefully?

Reply:
> The pull subscription model uses an event object (see the [CreateEventEx](https://learn.microsoft.com/en-us/windows/desktop/api/synchapi/nf-synchapi-createeventexa) function) to signal the application that there are events in the result set that match the query criteria.

This means that the EvtSubscribe should update the state of signal event. 

>  Create a loop construct that waits on the event object until the event is signaled.

This could be replaced with the waiting yield loop in our implementation:
https://github.com/fluent-plugins-nursery/winevt_c/blob/master/ext/winevt/winevt_subscribe.c#L475

> Then, call the [EvtNext](https://learn.microsoft.com/en-us/windows/desktop/api/WinEvt/nf-winevt-evtnext) function in a loop to enumerate the events in the result set. When the EvtNext function fails and sets the last error to ERROR_NO_MORE_ITEMS, reset the event object and wait for the service to signal the object again when there are events in the result set.

This is definitely I impemented this PR:
https://github.com/fluent-plugins-nursery/winevt_c/pull/44/files#diff-a0c0ab963ceb07088766ba81157d068e92b869e88b66331018121c73a4a46dd6

Reply:
Thanks!

I wasn't sure what the purpose of using this signal was, but in the official example, it is used to wait for new data.

`rb_winevt_subscribe_next` is a function that enumerates all the current data, right?
Then, I don't see the need to wait for the signal as in this example.

In our case, we want to use this signal to detect errors, right?

Reply:
> `rb_winevt_subscribe_next` is a function that enumerates all the current data, right?

Strictly, `rb_winevt_subscribe_each`.

Reply:
> > `rb_winevt_subscribe_next` is a function that enumerates all the current data, right?
> 
> Strictly, `rb_winevt_subscribe_each`.

The main function of yielding the collected data is: https://github.com/fluent-plugins-nursery/winevt_c/blob/d8299fbfbd723c05f43fb0f84cbf9da255058757/ext/winevt/winevt_subscribe.c#L452

Reply:
> > This implementation is inspired from the article: https://learn.microsoft.com/ja-jp/windows/win32/wes/subscribing-to-events#pull-subscriptions EvtSubscribe can tell internal state with a registered singal event.
> 
> That's the what I wanted to know, thanks! Could you add this link as a comment or commit message? (language should be en-us)
> 
> > When using INFINITE timeout here, it will be waiting forever even if SIGINT is caught in the parent process of Fluentd. It's sick behavior for us.
> 
> I agree with you, I didn't intend to use `INFINITE` here. But how about using small non zero value (e.g. 100) instead?

I misunderstood it.
If it need to wait, it means there is no more event to process.
In this case, it should return immediately and check trailing events in next call.
The timer will be processed in callee, such as Fluend's event loop.
Now I've understand that the value should be `0` here.

Reply:
I agree that there is no need to wait here.

Reply:
Yup. We always wait and executed by timer loops in Fluentd. This gem just provides abilities to call Windows native APIs. 
In short, we don't have to wait here with above zero milliseconds.